### ISSUE TITLE
[Feat]  EncoderModelRunner: shared encoder runner template

### ISSUE THREAD

**Issue by JingwenGu0829** [sgl-project/sglang-omni#512]
## Motivation

Issue #493 tracks reusable templates for common Omni optimizations. Encoder execution is a good first subissue because image, video, and audio encoders already share a clear stage lifecycle, while their optimization logic is currently implemented inside model-specific stage factories.

Today Qwen3-Omni has encoder batching, cache lookup/store, duplicate cache-key dedup, output splitting, and skip handling in `qwen3_omni/stages.py`. Ming-Omni has separate image/audio encoder closures with the same stage IO shape. As more Omni models are added, this lifecycle should become a clean runner surface that model developers can adapt by implementing the tensor-specific parts.

The proposed abstraction is a small `EncoderModelRunner` with an explicit:

```text
pre -> forward -> post
```

flow. The runner owns the common stage mechanics. Each model runner owns its own tensor semantics.

## Current encoder mechanism

An encoder stage is a single-pass model stage:

1. Preprocessing writes `PipelineState.encoder_inputs[stage_name]`.
2. The encoder stage reads its stage input.
3. It handles `_skip` requests and optional `cache_key`.
4. It normalizes and batches model inputs.
5. It runs the model forward.
6. It splits model outputs back to each request.
7. It writes `PipelineState.encoder_outs[stage_name]` and `engine_outputs[stage_name]`.
8. The aggregate stage merges encoder outputs into thinker `model_inputs`.

The common lifecycle is stable, while the tensor rules differ by model and modality:

- vision encoders use patch tensors plus `grid_thw` metadata;
- video often shares the image tower but uses separate input/output keys;
- audio encoders use feature tensors plus masks or lengths;
- some encoders concatenate packed items;
- some encoders pad by time;
- some encoders unwrap and rewrap segments;
- some outputs include only embeddings;
- some outputs include grids, token counts, output lengths, or deepstack tensors.

## Proposed design

Add an `EncoderModelRunner` for single-pass encoder stages:

```python
class EncoderModelRunner:
    def execute(self, payloads: list[StagePayload]) -> list[StagePayload]:
        prepared = self.pre(payloads)
        raw_outputs = self.forward(prepared)
        return self.post(prepared, raw_outputs)
```

The base runner should be small. It should make the lifecycle clear and provide shared helpers for the mechanics that are already common:

- stage input parsing;
- `_skip` result propagation;
- cache lookup/store;
- duplicate cache-key dedup inside one batch;
- common no-grad or inference-mode boundary;
- graph hit/miss and eager fallback stats;
- payload result writeback.

Model-specific runners should implement the tensor-specific pieces:

- how to count items;
- how to compute per-item output tokens;
- how to slice a sub-batch;
- how to concatenate or pad inputs;
- how to prepare CUDA graph buffers;
- how to run eager forward;
- how to run graph forward;
- how to split and clone outputs.

This gives future model authors a small set of methods to adapt, while keeping the reusable backend clean.

## Phase methods

The runner should group customization by phase. The method names below are illustrative; the important part is the phase boundary.

### Pre

`pre` turns stage payloads into a prepared encoder batch. It can use shared helpers for cache and skip, then call model-specific hooks for shape planning.

Common work:

- load `PipelineState`;
- build `EncoderRequest` objects;
- handle `_skip`;
- check cache;
- dedup duplicate cache keys in the same incoming batch;
- filter active cache misses;
- compute per-item metadata;
- plan eager or graph sub-batches;
- prepare replay buffers for graph-capable sub-batches.

Model-specific hooks: 

```python
def build_encoder_request(...)
def prepare_batch(...)
def estimate_payload_cost(...)
def encoder_eager_forward(...)
def encoder_cudagraph_forward(...)
def split_output(...)
def apply_encoder_result(...)
```

### Forward

`forward` receives a prepared batch or prepared sub-batches and runs either graph replay or eager fallback.

Common work:

- select the smallest fitting graph budget;
- copy current tensors into stable graph buffers;
- replay the graph;
- fall back to eager for oversized or unsupported inputs;
- count graph hits and misses.

Model-specific hooks:

```python
def encoder_eager_forward(self, batch: EncoderBatch) -> Any: ...
def encoder_cudagraph_forward(
    self,
    graph_inputs: EncoderGraphInputs,
    buffers: dict[str, torch.Tensor],
) -> Any: ...
```

### Post

`post` turns raw model outputs into per-request stage outputs.

Common work:

- restore original request order after item packing;
- clone graph outputs before graph buffers are reused;
- store cacheable results;
- attach deduplicated cache results to duplicate waiters;
- write `encoder_outs[stage_name]`;
- write `engine_outputs[stage_name]`;
- return updated payloads.

Model-specific hooks:

```python
def split_output(
    self,
    raw_output: Any,
    prepared_batch: PreparedEncoderBatch,
) -> dict[int, Any]: ...

def apply_encoder_result(
    self,
    payload: StagePayload,
    result: Any,
) -> StagePayload: ...
```

## Qwen3-Omni target

The first implementation target should be Qwen3-Omni image/video encoder.

Reasons:

- image and video share the same vision tower;
- current code already has batching and output splitting;
- current code already has cache and duplicate-key dedup;
- output token counts are available from `grid_thw` and merge size;
- graph budgets can naturally be based on visual output tokens;
- eager fallback preserves behavior for oversized media.

The Qwen vision runner should preserve current output keys:

- `image_embeds`
- `image_grid_thw`
- `image_token_counts`
- `deepstack_visual_embeds_image`
- `video_embeds`
- `video_grid_thw`
- `video_token_counts`
- `deepstack_visual_embeds_video`

The Qwen audio runner can follow after the image/video runner lands. It should share the same base runner lifecycle while using audio-specific pre/post:

- derive lengths from `audio_feature_lengths` or `feature_attention_mask`;
- pad feature tensors to max time;
- forward through `Qwen3OmniAudioEncoder`;
- split embeddings by `audio_output_lengths`.

**Comment by nagisa-kunhah** [comment]
@JingwenGu0829 Hello, can I give this a try?
I've looked through the related code. Here's what I'll do:

1. Create `sglang_omni/encoder_runner/base.py` with an `EncoderModelRunner` base class that owns the common `pre → forward → post` template:
   - **pre**: Parse stage input from `PipelineState`, handle `_skip` requests, LRU cache lookup, same-batch duplicate cache-key dedup, filter active (cache-miss) items into a batch-ready list.
   - **forward**: Concatenate/pad tensors into a single batched input, run `model(**batched_inputs)` under `torch.no_grad()`.
   - **post**: Split batched model output back to per-request results, store results into LRU cache, resolve duplicate waiters with the leader's result, write `encoder_outs`/`engine_outputs` into `PipelineState`.

   The base class implements:
   - `execute_single(payload)` — full single-request path (skip → cache → forward → store → apply).
   - `execute_batch(payloads)` — the template method orchestrating pre/forward/post for a batch.
   - `_lookup_cache()` / `_store_cache()` — LRU cache operations.

   And defines abstract hooks for subclasses:
   - `is_batchable(request)` — whether a request's tensors can be batched.
   - `prepare_batch(active_items)` — model-specific tensor concatenation/padding.
   - `encoder_forward(batched_inputs)` — the actual model call.
   - `split_output(raw_output, batch_meta)` — slice batched output back to per-request dicts.

2. Create `sglang_omni/models/qwen3_omni/encoder_runners.py` with two subclasses derived from `EncoderModelRunner` — `Qwen3VisionEncoderRunner` and `Qwen3AudioEncoderRunner` — that implement the four hooks above with Qwen3-specific tensor logic.

3. Update the factory functions (`create_image_encoder_executor`, `create_audio_encoder_executor`) in `stages.py` to instantiate the runner and pass `runner.execute_batch` as `batch_compute_fn` to `SimpleScheduler`.

**Comment by JingwenGu0829** [comment]
Hi @nagisa-kunhah Sure, we can do it together if you want, since it envolves some further cleanup in general model runner I feel essential to include in this pr too. So I think you can go ahead with the plan, start a PR and we can refine it together!

**Comment by JingwenGu0829** [comment]
Hi @nagisa-kunhah This task is taken by some other members and I already. Would you mind looking for some issues, especially those which are more open to first-time contributors? For exmaple https://github.com/sgl-project/sglang-omni/pull/451#issuecomment-4521431968 (someone may have taken this issue too, but this type of issue is more open to first time contributor rather than this type of framework-level work).

**Comment by nagisa-kunhah** [comment]
@JingwenGu0829 Got it, I'll look for other issues.

**Comment by Ccyest** [comment]
Heads up: PR #605 implements a reference-audio cache for Higgs TTS (waveform cache in preprocessing + delayed codec-code cache in the encoder, plus stable cache-key derivation from path/bytes/base64).

When building the `EncoderModelRunner` here, worth considering two things from #605 that this issue doesn't currently cover:
1. **Where the cache_key comes from** — #605 derives a stable key from the raw media input (path uses size+mtime_ns to avoid stale hits). This issue assumes cache_key already exists.
2. **Preprocessing-stage caching** — #605 also caches the loaded/resampled waveform in preprocessing, which is outside the encoder stage this runner targets.

The encoder-side code cache in #605 should fold cleanly into this runner; the key-derivation + preprocessing cache are complementary. FishAudio S2 Pro (`fishaudio_s2_pro/stages.py`) does the same load→resample→encode on reference audio with no cache, so it'd be a good second adopter.

**Comment by JingwenGu0829** [comment]
@Ccyest Yes thanks for the reminder!
